compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -268,7 +268,20 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
             if (qualifies(defDenot)) {
               val found =
                 if (isSelfDenot(defDenot)) curOwner.enclosingClass.thisType
-                else curOwner.thisType.select(name, defDenot)
+                else {
+                  val effectiveOwner =
+                    if (curOwner.isTerm && defDenot.symbol.isType)
+                      // Don't mix NoPrefix and thisType prefixes, since type comparer
+                      // would not detect types to be compatible. Note: If we replace the
```

Comment:
It's only in the applied type branch that this became an issue. We saw the same type parameter TypeRef once with a NoPrefix prefix, the other with a ThisType. They should be the same but subtype checking failed. We could make subtype checking more lenient, but I think it's better not to have variations how we construct these things. I believe the current scheme hides the difference because we follow an alias from TypeParamAccessor to TypeParam, that's why we did not see it so far.